nomad/event_sink_manager.go

```diff
@@ -0,0 +1,565 @@
+	// SinkManager should stop
+	ctx context.Context
+
+	// updateSinkInterval is the interval to update event sink progress to raft
+	updateSinkInterval time.Duration
+
+	// mu synchronizes access to sinkSubscriptions and eventSinksWs, running
+	// and shutdownCh
+	mu sync.Mutex
+
+	// sinkSubscriptions are the set of managed sinks that the manager is tracking
+	sinkSubscriptions map[string]*ManagedSink
+
+	// eventSinksWs is a watchset used to check for new event sinks added to the state store
+	eventSinksWs memdb.WatchSet
+
+	// running specifies if the manager is running
+	running bool
+
+	// shutdownCh is used to stop the manager from running
+	shutdownCh chan struct{}
+
+	// delegate is the interface needed to interact with State and RPCs
+	delegate SinkDelegate
+
+	L hclog.Logger
+}
+
+// SinkDelegate is the interface needed for the SinkManger to interfact with
+// parts of Nomad
+type SinkDelegate interface {
+	State() *state.StateStore
+	getLeaderAcl() string
+	Region() string
+	RPC(method string, args interface{}, reply interface{}) error
+}
+
+type serverDelegate struct {
+	*Server
+}
+
+// NewSinkManager builds a new SinkManager. It also creates ManagedSinks for
+// all EventSinks in the state store
+func NewSinkManager(ctx context.Context, delegate SinkDelegate, l hclog.Logger) *SinkManager {
+	m := &SinkManager{
+		ctx:                ctx,
+		delegate:           delegate,
+		updateSinkInterval: 30 * time.Second,
+		sinkSubscriptions:  make(map[string]*ManagedSink),
+		L:                  l.Named("sink_manager"),
+	}
+
+	return m
+}
+
+// EstablishManagedSinks creates and sets ManagedSinks for the Manager
+func (m *SinkManager) EstablishManagedSinks() error {
+	m.eventSinksWs = m.delegate.State().NewWatchSet()
+
+	iter, err := m.delegate.State().EventSinks(m.eventSinksWs)
+	if err != nil {
+		return err
+	}
+
+	var sinkIDs []string
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		sink := raw.(*structs.EventSink)
+		sinkIDs = append(sinkIDs, sink.ID)
+	}
+
+	for _, id := range sinkIDs {
+		mSink, err := NewManagedSink(m.ctx, id, m.delegate.State, m.L)
+		if err != nil {
+			return fmt.Errorf("creating managed sink: %w", err)
+		}
+		m.sinkSubscriptions[id] = mSink
+	}
+	return nil
+}
+
+// Running specifies if the manager is currently running
+func (m *SinkManager) Running() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.running
+}
+
+// Stop stops the manager
+func (m *SinkManager) Stop() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if !m.running {
+		return
+	}
+	close(m.shutdownCh)
+	m.running = false
+}
+
+// Run is a long running function that starts all of the ManagedSinks.
+func (m *SinkManager) Run() error {
+	m.mu.Lock()
+	m.running = true
+	m.shutdownCh = make(chan struct{})
+	m.mu.Unlock()
+
+	errCh := make(chan SinkError)
+
+	updateSinks := time.NewTicker(m.updateSinkInterval)
+
+	run := func(id string, ms *ManagedSink) {
+		err := ms.Run()
+		select {
+		case <-m.ctx.Done():
```

Comment:
If we don't pass this context into the `ms.Run()`, what does waiting for `Done()` do for us here? We don't that the context is done until `ms.Run()` returns, and then we'd end up racing the select between the context and the error channel. I _think_ you could fix this (and generally make things a little more legible) by passing the `m.ctx` down into the `ManagedSink` and having the managed sink select on that context in its loop.

Reply:
I think passing in the context to managed sink's run is cleaner, its not obvious that the same contex was set in the managed sink's constructor, so passing it in here helps with that.

We still need the select ctx.Done, below as well, that way we can either send an error to the manager to log/ and deregister the sink if it was deleted, or skip that step if we are shutting down